mods/carts/cart_entity.lua

```diff
@@ -173,17 +173,17 @@ local function rail_on_step(self, dtime)
 		return
 	end
 
+	local pos = self.object:getpos()
+	local update = {}
+
 	-- stop cart if velocity vector flips
-	if self.old_vel and (((self.old_vel.x * vel.x) < 0) or
-			((self.old_vel.z * vel.z) < 0)) and
-			(self.old_vel.y == 0) then
-		self.old_dir = {x = 0, y = 0, z = 0}
+	if self.old_vel and self.old_vel.y == 0 and
+			(self.old_vel.x * vel.x < 0 or self.old_vel.z * vel.z < 0) then
 		self.old_vel = {x = 0, y = 0, z = 0}
-		self.velocity = {x = 0, y = 0, z = 0}
```

Comment:
Make a long track of brake rails. Punch the cart around. Once in a while it reverses almost an entire node and then jerks back. With that line put back, I don't seem to have that problem.